{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "16bf9d78",
   "metadata": {},
   "source": [
    "# Example of Metadata Tracking with OpenCGA\n",
    "OpenCGA has many components, one of the most important for this project being OpenCGA Catalog. This component allows users to create a metadata database to track users, projects, studies, files, samples, families, and jobs. We will interact with this database using the Python client library `PyOpenCGA` to access the REST API included in OpenCGA. You can refer to the documentation [here](http://docs.opencb.org/display/opencga/Python) to learn more about PyOpenCGA. \n",
    "\n",
    "This notebook will act as an example for how to use OpenCGA to track metadata in the Silent Genomes Project. More specifically, we will include examples showing how to use `PyOpenCGA` with the OpenCGA instance created on the BCCHR server.\n",
    "\n",
    "The following links can provide further reference for describing OpenCGA and how to use it:\n",
    "* [Overview of Data Management](http://docs.opencb.org/display/opencga/Data+Management) including definitions of the major entities involved in OpenCGA Catalog"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77abbf9e",
   "metadata": {},
   "source": [
    "# Imports\n",
    "The following cell imports the libraries needed to run code within this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9f914786",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries needed to run the script\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from time import sleep\n",
    "from pprint import pprint\n",
    "\n",
    "from result import *\n",
    "from pyopencga.opencga_config import ClientConfiguration\n",
    "from pyopencga.opencga_client import OpenCGAClient"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb92af4",
   "metadata": {},
   "source": [
    "# Configure and Connect\n",
    "Before using OpenCGA, the client needs to be configured and connected to. Information relating to the OpenCGA host, username and password are considered to be sensitive information and should not be openly shared. For this reason, they are stored as environment variables in a `.env` file located in the base directory of this project. \n",
    "\n",
    "In order to read the information from this file into your environment, before running this notebook you need to activate your environment and load the variables with the following command:\n",
    "```\n",
    "source venv/bin/activate\n",
    "source .env\n",
    "```\n",
    "\n",
    "This should load the variables `OPENCGA_HOST`, `OPENCGA_USERNAME`, and `OPENCGA_PASSWORD` into your environment so you can successfully connect to the client."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a0cd5911",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure the OpenCGA client\n",
    "config = ClientConfiguration({\n",
    "        \"rest\": {\n",
    "                \"host\": os.environ.get(\"OPENCGA_HOST\")\n",
    "        }\n",
    "})\n",
    "oc = OpenCGAClient(config)\n",
    "\n",
    "# Authenticate the user\n",
    "user = os.environ.get(\"OPENCGA_USERNAME\")\n",
    "passwd = os.environ.get(\"OPENCGA_PASSWORD\")\n",
    "oc.login(user, passwd)\n",
    "\n",
    "# Setup OpenCGA Clients\n",
    "users = oc.users\n",
    "projects = oc.projects\n",
    "studies = oc.studies\n",
    "files = oc.files\n",
    "jobs = oc.jobs\n",
    "families = oc.families\n",
    "individuals = oc.individuals\n",
    "samples = oc.samples\n",
    "cohorts = oc.cohorts\n",
    "panels = oc.panels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09661f2c",
   "metadata": {},
   "source": [
    "# Querying the Metadata Database\n",
    "The following section outlines a number of queries that can be performed using OpenCGA. Information in this database is organized into the following major entities:\n",
    "1. **_Users_**: a person who will be using OpenCGA -- every user should be authenticated and some can perform specific actions\n",
    "2. **_Groups_**: you can create a group of users to simplify data permission management. These are defined at the study level (i.e. each study contains different groups) and can only be created by the study owner\n",
    "3. **_Projects_**: a piece of planned work or an activity that is finished over a period of time and is intended to achieve a particular purpose. Any user with full permissions can create any number of projects and studies. Projects need to contain at minimum, a name, an alias (project identifier), and the species organism\n",
    "4. **_Studies_**: projects are componsed by a set of studies. A study is the activity of examining a subject in detail in order to discover new information. Aside from Users and Projects, most of the data models in Catalog belong to a particular study\n",
    "5. **_Variable Sets_**: a set of Variables (the complete definition of fields that need to be populated). Every Study can have as many different Variable Set definitions as necessary\n",
    "6. **_Annotation Sets_**: the values defined for each of the Variables are called Annotations, and the population of a whole Variable Set is called an Annotation Set. An Annotation Set is always related to one concrete Variable Set\n",
    "7. **_Files_**: every File registry contains the physical path where the files/folders are stored in the file system (i.e. the uri). Catalog also creates a virtual file structure so no matter what the real location of the files are, users can organize and work with those files differently\n",
    "8. **_Individuals and Families_**: an Individual is a subject (typically a person) for which some analysis will be made. A group of Individuals with any parental or blood relationship is called a Family\n",
    "9. **_Samples and Cohorts_**: a Sample is any biological material, normally extracted from an Individual. Cohorts contain groups of samples sharing some particular conditions such as \"healthy\" vs \"infected\"\n",
    "10. **_Clinical Analysis_**: contains all the information of the Individuals and Samples involved to perform a real clinical analysis\n",
    "11. **_Job_**: every time the user calls to an analysis webservice to run anything, a new Job is created. This job contains the essential information about the task that needs to be run. OpenCGA supports SGE\n",
    "\n",
    "The above entities are related in the following way:\n",
    "![OpenCGA Data Model](http://docs.opencb.org/download/attachments/327907/catalog_data_models_v13.png?version=1&modificationDate=1560245879990&api=v2)\n",
    "\n",
    "\n",
    "For more information regarding API endpoints for OpenCGA, you can refer to the demo Swagger located [here](http://bioinfo.hpc.cam.ac.uk/opencga-demo/webservices/#/). \n",
    "\n",
    "If you need more information about an object or method, you can use the `help()` command as shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c3813dc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Samples in module pyopencga.rest_clients.sample_client object:\n",
      "\n",
      "class Samples(pyopencga.rest_clients._parent_rest_clients._ParentBasicCRUDClient, pyopencga.rest_clients._parent_rest_clients._ParentAclRestClient, pyopencga.rest_clients._parent_rest_clients._ParentAnnotationSetRestClient)\n",
      " |  Samples(configuration, session_id=None, login_handler=None, *args, **kwargs)\n",
      " |  \n",
      " |  This class contains method for Samples webservice\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Samples\n",
      " |      pyopencga.rest_clients._parent_rest_clients._ParentBasicCRUDClient\n",
      " |      pyopencga.rest_clients._parent_rest_clients._ParentAclRestClient\n",
      " |      pyopencga.rest_clients._parent_rest_clients._ParentAnnotationSetRestClient\n",
      " |      pyopencga.rest_clients._parent_rest_clients._ParentRestClient\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, configuration, session_id=None, login_handler=None, *args, **kwargs)\n",
      " |      :param login_handler: a parameterless method that can log in this connector\n",
      " |      and return a session id\n",
      " |  \n",
      " |  aggregation_stats(self, **options)\n",
      " |      Fetch catalog sample stats\n",
      " |      URL: /{apiVersion}/samples/aggregationStats\n",
      " |      \n",
      " |      :param study: study [[user@]project:]study where study and project can be either the id or alias\n",
      " |      :param source: Source\n",
      " |      :param creationYear: Creation year\n",
      " |      :param creationMonth: Creation month (JANUARY, FEBRUARY...)\n",
      " |      :param creationDay: Creation day\n",
      " |      :param creationDayOfWeek: Creation day of week (MONDAY, TUESDAY...)\n",
      " |      :param status: Status\n",
      " |      :param type: Type\n",
      " |      :param phenotypes: Phenotypes\n",
      " |      :param release: Release\n",
      " |      :param version: Version\n",
      " |      :param somatic: Somatic (Boolean, True/False)\n",
      " |      :param annotation: Annotation, e.g: key1=value(;key2=value)\n",
      " |      :param default: Calculate default stats (Boolean, False as default)\n",
      " |      :param field: List of fields separated by semicolons, e.g.: studies;type. For nested\n",
      " |          fields use >>, e.g.: studies>>biotype;type;numSamples[0..10]:1\n",
      " |  \n",
      " |  load(self, file, **options)\n",
      " |      Load samples from a ped file [EXPERIMENTAL]\n",
      " |      URL: /{apiVersion}/samples/load\n",
      " |      \n",
      " |      :param study: Study [[user@]project:]study where study and project can be either the id or alias\n",
      " |      :param file: Filename of study linked file\n",
      " |      :param variableSet: variableSet\n",
      " |  \n",
      " |  search(self, **options)\n",
      " |      Sample search method\n",
      " |      URL: /{apiVersion}/samples/search\n",
      " |      \n",
      " |      :param study: Study [[user@]project:]{study1,study2|*} where studies and project can be either the id or alias.\n",
      " |      :param name: [DEPRECATED] name\n",
      " |      :param source: source\n",
      " |      :param type: type\n",
      " |      :param somatic: boolean (True, False)\n",
      " |      :param individual: Individual id or name\n",
      " |      :param creationDate: Creation date (Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805...)\n",
      " |      :param modificationDate: Modification date (Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805...)\n",
      " |      :param phenotypes: Comma separated list of phenotype ids or names\n",
      " |      :param annotationsetName: DEPRECATED: Use annotation queryParam this way: annotationSet[=|==|!|!=]{annotationSetName}\n",
      " |      :param variableSet: [DEPRECATED] Use annotation queryParam this way: variableSet[=|==|!|!=]{variableSetId}\n",
      " |      :param annotation: Annotation, e.g: key1=value(;key2=value)\n",
      " |      :param attributes: Text attributes (Format: sex=male,age>20 ...)\n",
      " |      :param nattributes: Numerical attributes (Format: sex=male,age>20 ...)\n",
      " |      :param skipCount: Skip count (Boolean, True as default)\n",
      " |      :param release: Release value (Current release from the moment the samples were first created)\n",
      " |      :param snapshot: Snapshot value (Latest version of samples in the specified release)\n",
      " |      :param include: Fields included in the response, whole JSON path must be provided\n",
      " |      :param exclude: Fields excluded in the response, whole JSON path must be provided\n",
      " |      :param limit: Number of results to be returned in the queries\n",
      " |      :param skip: Number of results to skip in the queries\n",
      " |      :param count: Total number of results (Boolean, False as default)\n",
      " |      :param includeIndividual: Include Individual object as an attribute (this replaces old lazy parameter) (Boolean, False as default)\n",
      " |      :param flattenAnnotations: Flatten the annotations? (Boolean, False as default)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pyopencga.rest_clients._parent_rest_clients._ParentBasicCRUDClient:\n",
      " |  \n",
      " |  create(self, data, **options)\n",
      " |  \n",
      " |  delete(self, **options)\n",
      " |  \n",
      " |  info(self, query_id, **options)\n",
      " |  \n",
      " |  update(self, query_id, data, **options)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pyopencga.rest_clients._parent_rest_clients._ParentAclRestClient:\n",
      " |  \n",
      " |  acl(self, query_id, **options)\n",
      " |      acl info\n",
      " |      \n",
      " |      :param query_id:\n",
      " |      :param options:\n",
      " |  \n",
      " |  update_acl(self, memberId, data, **options)\n",
      " |      update acl\n",
      " |      \n",
      " |      :param query_id:\n",
      " |      :param options:\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pyopencga.rest_clients._parent_rest_clients._ParentAnnotationSetRestClient:\n",
      " |  \n",
      " |  update_annotations(self, query_id, annotationset_id, data, **options)\n",
      " |      update annotations from an AnnotationSet\n",
      " |      \n",
      " |      :param query_id: Entry identifier.\n",
      " |      :param annotationset_id: AnnotationSet id.\n",
      " |      :param data: Json containing the map of annotations when the action is ADD, SET or REPLACE, a json with only\n",
      " |      the key 'remove' containing the comma separated variables to be removed as a value when the action is REMOVE\n",
      " |      or a json with only the key 'reset' containing the comma separated variables that will be set to the default\n",
      " |      value when the action is RESET\n",
      " |      :param options: QueryParam options\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pyopencga.rest_clients._parent_rest_clients._ParentRestClient:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(oc.samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd3887da",
   "metadata": {},
   "source": [
    "### Querying for Samples\n",
    "Samples can be organized into projects and/or studies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a56b0e68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1000"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.search(study='1kG_phase3').num_results()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "01c9f419",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'annotationSets': [],\n",
      " 'attributes': {'OPENCGA_INDIVIDUAL': {'affectationStatus': 'UNKNOWN',\n",
      "                                       'attributes': {},\n",
      "                                       'creationDate': '20210901022141',\n",
      "                                       'disorders': [],\n",
      "                                       'ethnicity': 'AFR',\n",
      "                                       'father': {'parentalConsanguinity': False,\n",
      "                                                  'release': 0,\n",
      "                                                  'version': 0},\n",
      "                                       'id': 'HG00096',\n",
      "                                       'karyotypicSex': 'XY',\n",
      "                                       'lifeStatus': 'ALIVE',\n",
      "                                       'location': {},\n",
      "                                       'modificationDate': '20210904233501',\n",
      "                                       'mother': {'parentalConsanguinity': False,\n",
      "                                                  'release': 0,\n",
      "                                                  'version': 0},\n",
      "                                       'multiples': {},\n",
      "                                       'name': 'HG00096',\n",
      "                                       'parentalConsanguinity': False,\n",
      "                                       'phenotypes': [],\n",
      "                                       'population': {'description': '',\n",
      "                                                      'name': '',\n",
      "                                                      'subpopulation': ''},\n",
      "                                       'release': 1,\n",
      "                                       'sex': 'MALE',\n",
      "                                       'status': {'date': '20210901022141',\n",
      "                                                  'message': '',\n",
      "                                                  'name': 'READY'},\n",
      "                                       'uuid': 'nyl7KAF7AAYAAacraMMiSg',\n",
      "                                       'version': 1}},\n",
      " 'creationDate': '20210824224952',\n",
      " 'description': 'demo description',\n",
      " 'id': 'HG00096',\n",
      " 'individual': {'parentalConsanguinity': False, 'release': 0, 'version': 0},\n",
      " 'modificationDate': '20211109220553',\n",
      " 'phenotypes': [],\n",
      " 'release': 1,\n",
      " 'somatic': False,\n",
      " 'source': 'ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz',\n",
      " 'stats': {},\n",
      " 'status': {'date': '20210824224952', 'message': '', 'name': 'READY'},\n",
      " 'type': '',\n",
      " 'uuid': 'elsLEwF7AAQAAbVUxt8mYA',\n",
      " 'version': 1}\n",
      "{'annotationSets': [],\n",
      " 'attributes': {},\n",
      " 'creationDate': '20210824224952',\n",
      " 'description': '',\n",
      " 'id': 'HG00097',\n",
      " 'individual': {'parentalConsanguinity': False, 'release': 0, 'version': 0},\n",
      " 'phenotypes': [],\n",
      " 'release': 1,\n",
      " 'somatic': False,\n",
      " 'source': 'ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz',\n",
      " 'stats': {},\n",
      " 'status': {'date': '20210824224952', 'message': '', 'name': 'READY'},\n",
      " 'type': '',\n",
      " 'uuid': 'elsLPgF7AAQAAXExYzGpFg',\n",
      " 'version': 1}\n"
     ]
    }
   ],
   "source": [
    "sample_results = samples.search(study='1kG_phase3', limit=2)\n",
    "\n",
    "for res in sample_results.results():\n",
    "    pprint(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ff5da176",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on RESTResponse in module pyopencga.commons object:\n",
      "\n",
      "class RESTResponse(builtins.object)\n",
      " |  RESTResponse(response)\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, response)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  num_matches(self)\n",
      " |      Return the total number of matches taking of all the DataResponses\n",
      " |  \n",
      " |  num_results(self)\n",
      " |      Return the total number of results taking of all the DataResponses\n",
      " |  \n",
      " |  results(self)\n",
      " |      Iterates over all the results of all the QueryResults\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(samples.search())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bf08116",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67b36d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73920762",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4850351",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bacdbff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48dacc9a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
